document.go

```diff
@@ -356,6 +337,52 @@ func (d Document) Flag(flag DocumentFlag) bool {
 	return d.data.flag.Is(flag)
 }
 
+// Get field by index. Flag is false if this field is inaccessible
+func (d Document) fieldByIndex(index []int, init bool) (reflect.Value, bool) {
```

Comment:
how about modifying original implementation of [FieldByIndex](https://cs.opensource.google/go/go/+/refs/tags/go1.17.7:src/reflect/value.go;bpv=1;bpt=1;l=1228?gsn=FieldByIndex&gs=kythe%3A%2F%2Fgolang.org%3Flang%3Dgo%3Fpath%3Dreflect%23method%2520Value.FieldByIndex) to add logic for lazy initialization?

I'm thinking something like this:
```golang
func reflectValueFieldByIndex(v reflect.Value, index []int, init bool) Value {
	if len(index) == 1 {
		return v.Field(index[0])
	}
	if v.Kind() != reflect.Struct {
		panic("rel: value must be a struct")
	}
	for i, x := range index {
		if i > 0 {
			if v.Kind() == Ptr && v.Type().Elem().Kind() == Struct {
				if v.IsNil() {
					if !init {
						return false
					}
					field.Set(reflect.New(v.Type().Elem()))
				}
				v = v.Elem()
			}
		}
		v = v.Field(x)
	}
	return v
}
```

Reply:
I've merged it with checkPointersForIndices, so yes, it doesn't walk down the types a second time